src/IO/ADIOS/ADIOS2IOHandler.cpp

```diff
@@ -20,58 +20,1236 @@
+    {
+        throw std::runtime_error( "Creating a dataset in a file opened as read "
+                                  "only is not possible." );
+    }
+    if ( !writable->written )
+    {
+        /* Sanitize name */
+        std::string name = auxiliary::removeSlashes( parameters.name );
+
+        auto file = refreshFileFromParent( writable );
+        auto filePos = setAndGetFilePosition( writable, name );
+        filePos->gd = ADIOS2FilePosition::GD::DATASET;
+        auto varName = filePositionToString( filePos );
+        // we use a unique_ptr to circumvent the fact that std::optional
+        // is only available beginning with c++17
+        std::unique_ptr< adios2::Operator > compression;
+        if ( parameters.compression == "zfp" )
+        {
+            compression = std::unique_ptr< adios2::Operator >{
+                new adios2::Operator( this->m_zfp )};
+        }
+        else if ( parameters.compression == "sz" )
+        {
+            compression = std::unique_ptr< adios2::Operator >{
+                new adios2::Operator( this->m_sz )};
+        }
+        switchType( parameters.dtype, detail::VariableDefiner( ),
+                    getFileData( file ).m_IO, varName,
+                    std::unique_ptr< adios2::Operator >( ), parameters.extent );
+        writable->written = true;
+        m_dirty.emplace( file );
+    }
+}
+
+void ADIOS2IOHandlerImpl::extendDataset(
+    Writable *, const Parameter< Operation::EXTEND_DATASET > & )
+{
+    throw std::runtime_error(
+        "Dataset extension not implemented in ADIOS backend" );
+}
+
+void ADIOS2IOHandlerImpl::openFile(
+    Writable * writable, const Parameter< Operation::OPEN_FILE > & parameters )
+{
+    if ( !auxiliary::directory_exists( m_handler->directory ) )
+    {
+        throw no_such_file_error( "Supplied directory is not valid: " +
+                                  m_handler->directory );
+    }
+
+    std::string name = parameters.name;
+    if ( !auxiliary::ends_with( name, ".bp" ) )
+    {
+        name += ".bp";
+    }
+
+    auto file = std::get< PE_InvalidatableFile >( getPossiblyExisting( name ) );
+
+    associateWithFile( writable, file );
+
+    writable->written = true;
+    writable->abstractFilePosition = std::make_shared< ADIOS2FilePosition >( );
+}
+
+void ADIOS2IOHandlerImpl::openPath(
+    Writable * writable, const Parameter< Operation::OPEN_PATH > & parameters )
+{
+    /* Sanitize path */
+    refreshFileFromParent( writable );
+    std::string prefix =
+        filePositionToString( setAndGetFilePosition( writable->parent ) );
+    std::string suffix = auxiliary::removeSlashes( parameters.path );
+    std::string infix = auxiliary::ends_with( prefix, '/' ) ? "" : "/";
+
+    /* ADIOS has no concept for explicitly creating paths.
+     * They are implicitly created with the paths of variables/attributes. */
+
+    if ( writable->abstractFilePosition )
+    {
+        *writable->abstractFilePosition = ADIOS2FilePosition(
```

Comment:
How awesome is that?! [Clang-tidy](https://travis-ci.org/openPMD/openPMD-api/jobs/535698753) found an issue of [object slicing](https://en.wikipedia.org/wiki/Object_slicing), cool!

If you really need to copy to the abstract class here, maybe consider implementing a `clone()` method or rewrite it another way.

Refs:
- https://github.com/openPMD/openPMD-api/blob/dev/include/openPMD/IO/IOTask.hpp / #410
- https://clang.llvm.org/extra/clang-tidy/checks/cppcoreguidelines-slicing.html